### DISCUSSION TITLE
Rerender only changed nodes

### DISCUSSION THREAD

**Discussion by Alan-Fisher** [xyflow/xyflow#3720] [Q&A]
Hey! How can I update (rerender) only changed nodes?

E.g. we change the text in one of the nodes and update only that one

Thank you for the great library btw

**Comment by moklick** [comment]
This should already the case. Are you using React Flow or Svelte Flow?

**Reply by Alan-Fisher** [reply]
Wow thanks for the rapid answer!

I use React Flow

The essential part of code looks like this:
```
useEffect(() => {
  const { nodes, edges } = jsonToFlow(json)
  setNodes(nodes)
  setEdges(edges)
}, [json])

return (
  <ReactFlow
    nodes={nodes}
    nodeTypes={nodeTypes}
    edges={edges}
    onlyRenderVisibleElements
  >
)
```

Maybe I should add some checks or special props?

**Reply by moklick** [reply]
Here you are replacing all nodes with the nodes from the json object. How do you want to update specific nodes?

**Reply by Alan-Fisher** [reply]
Yeah, my initial question was about this exact moment: how do I update only one node with reactflow API?

As I see, I can't use setNodes for a part of nodes

**Reply by Alan-Fisher** [reply]
@moklick :)

**Reply by alindsay55661** [reply]
I have the same question, I get changed nodes from an external source and only want to update those nodes as the graph is quite large. Yet, when any node changes, the entire graph is rerendered because I'm unsure how to selectively update...

**Reply by alindsay55661** [reply]
Or is this as simple as creating a single `nodes` array reference bound to `<ReactFlow>` and managing its contents (adding/removing/updating children manually) without overriding the array itself?

**Reply by moklick** [reply]
@Alan-Fisher you can update specific nodes, by passing a function:

```js
setNodes((nds) => nds.map((node) => {
  if (node.id === id) {
    return {
      ...node,
      position: {
       x: 100,
       y: 100
      }
    }
  }

  return node;
}));
```

we are working on a better setNodes for v12, but the API will be the same.
